File: frontend/src/api/gae.ts
import { request } from './request'
import type { ComputeGAEResponse, StepGAEResponse } from '@/types'

interface ComputeParams {
  tokens: string[]
  rewards: number[]
  values: number[]
  gamma?: number
  lambda?: number
}

interface StepParams {
  t: number
  tokens: string[]
  rewards: number[]
  values: number[]
  gamma?: number
  lambda?: number
  next_advantage?: number
}

export const gaeApi = {
  async compute(params: ComputeParams): Promise<ComputeGAEResponse> {
    return request<ComputeGAEResponse>({
      method: 'POST',
      url: '/api/gae/compute',
      data: {
        tokens: params.tokens,
        rewards: params.rewards,
        values: params.values,
        gamma: params.gamma ?? 0.99,
        lambda: params.lambda ?? 0.95,
      },
    })
  },

  async step(params: StepParams): Promise<StepGAEResponse> {
    return request<StepGAEResponse>({
      method: 'POST',
      url: '/api/gae/step',
      data: {
        t: params.t,
        tokens: params.tokens,
        rewards: params.rewards,
        values: params.values,
        gamma: params.gamma ?? 0.99,
        lambda: params.lambda ?? 0.95,
        next_advantage: params.next_advantage ?? 0.0,
      },
    })
  },
}

File: backend/app/api/gae.py
from fastapi import APIRouter, HTTPException
from app.models.common import ApiResponse
from app.models.request import ComputeGAERequest, StepGAERequest
from app.services.gae_service import compute_gae_full, step_gae

router = APIRouter()


@router.post("/compute", response_model=ApiResponse)
async def api_compute_gae(request: ComputeGAERequest):
    # 关联算法：C1(GAE完整计算) + 方差-偏差解读
    try:
        result = compute_gae_full(
            tokens=request.tokens,
            rewards=request.rewards,
            values=request.values,
            gamma=request.gamma,
            lam=request.lambda_,
        )
        return ApiResponse(success=True, data=result)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except ArithmeticError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/step", response_model=ApiResponse)
async def api_step_gae(request: StepGAERequest):
    # 关联算法：C2(单步GAE计算)
    try:
        result = step_gae(
            t=request.t,
            tokens=request.tokens,
            rewards=request.rewards,
            values=request.values,
            gamma=request.gamma,
            lam=request.lambda_,
            next_advantage=request.next_advantage,
        )
        return ApiResponse(success=True, data=result)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
